pilot/pkg/xds/endpoint_builder.go

```diff
@@ -370,3 +394,139 @@ func buildEnvoyLbEndpoint(e *model.IstioEndpoint) *endpoint.LbEndpoint {
 
 	return ep
 }
+
+// TODO this logic is probably done elsewhere in XDS, possible code-reuse + perf improvements
+type mtlsChecker struct {
+	push            *model.PushContext
+	svcPort         int
+	destinationRule *networkingapi.DestinationRule
+
+	// cache of host identifiers that have mTLS disabled
+	mtlsDisabledHosts map[string]struct{}
+
+	// cache of labels/port that have mTLS disabled by peerAuthn
+	peerAuthDisabledMTLS map[string]bool
+	// cache of labels that have mTLS disabled by a subset traffic policy
+	subsetPolicyDisabledMTLS map[string]bool
+	// true if the default traffic policy disables mTLS
+	disaledByDestinationRule bool
+}
+
+func newMtlsChecker(push *model.PushContext, svcPort int, dr *config.Config) *mtlsChecker {
+	var drSpec *networkingapi.DestinationRule
+	if dr != nil {
+		drSpec = dr.Spec.(*networkingapi.DestinationRule)
+	}
+	return &mtlsChecker{
+		push:                     push,
+		svcPort:                  svcPort,
+		destinationRule:          drSpec,
+		mtlsDisabledHosts:        map[string]struct{}{},
+		peerAuthDisabledMTLS:     map[string]bool{},
+		subsetPolicyDisabledMTLS: map[string]bool{},
+		disaledByDestinationRule: mtlsDisabledByDefaultTrafficPolicy(dr, svcPort),
+	}
+}
+
+// mTLSDisabled returns true if the given lbEp has mTLS disabled due to any of:
+// - disabled tlsMode
+// - DestinationRule disabling mTLS on the entire host or the port TODO handle subsets
+// - PeerAuthentication disabling mTLS at any applicable level (mesh, ns, workload, port)
+func (c *mtlsChecker) isMtlsDisabled(lbEp *endpoint.LbEndpoint) bool {
+	if c == nil {
+		return false
+	}
+	_, ok := c.mtlsDisabledHosts[lbEpKey(lbEp)]
+	return ok
+}
+
+// computeForEndpoint checks destination rule, peer authentication and metadata to determine if mTLS was turned off.
+// This must be done during conversion from IstioEndpoint since we still have workload metadata.
+func (c *mtlsChecker) computeForEndpoint(ep *model.IstioEndpoint) {
+	tlsMode := envoytransportSocketMetadata(ep.EnvoyEndpoint, model.TLSModeLabelShortname)
+	if tlsMode == model.DisabledTLSModeLabel ||
```

Comment:
are we sure this should be `tlsMode == model.DisabledTLSModeLabel` and not `tlsMode != model.IstioMutualTLSModeLabel`?

For example, look at `kubernetes.default` service - it will have no tlsMode.

I am confused since all tests pass though...

Reply:
You're right.. the tests are passing for some hidden reason. cx_connect_fail is incremented but we're not surfacing 500s

Reply:
hm.. that is a bit concerning. did you ever figure out why?

Reply:
```json
                    "route": {
                      "cluster": "outbound|5000||helloworld.naked.svc.cluster.local",
                      "timeout": "0s",
                      "retry_policy": {
                        "retry_on": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
                        "num_retries": 2,
                        "retry_host_predicate": [
                          {
                            "name": "envoy.retry_host_predicates.previous_hosts"
                          }
                        ],
                        "host_selection_retry_max_attempts": "5",
                        "retriable_status_codes": [
                          503
                        ]
                      },
```

I think it's the `retry_policy` causing this behavior.

Reply:
I thought we have the ExpectDestinations now that should catch it?